internals/stage_helpers.py

```diff
@@ -58,7 +76,91 @@ def get_feature_stage_ids_list(feature_id: int) -> list[dict[str, int]]:
       'intent_stage': INTENT_STAGES_BY_STAGE_TYPE.get(s.stage_type, INTENT_NONE)
     } for s in q]
 
+
 def get_ot_stage_extensions(ot_stage_id: int):
   """Return a list of extension stages associated with a stage in JSON format"""
   q = Stage.query(Stage.ot_stage_id == ot_stage_id)
   return [converters.stage_to_json_dict(stage) for stage in q]
+
+
+def get_stage_info_for_templates(
+    fe: FeatureEntry) -> StageTemplateInfo:
+  """Gather the information needed to display the estimated milestones table."""
+  # Only milestones from DevTrial, OT, or shipping stages are displayed.
+  id = fe.key.integer_id()
+  f_type = fe.feature_type or 0
+  proto_stage_type = STAGE_TYPES_PROTOTYPE[f_type]
+  dt_stage_type = STAGE_TYPES_DEV_TRIAL[f_type]
+  ot_stage_type = STAGE_TYPES_ORIGIN_TRIAL[f_type]
+  extension_stage_type = STAGE_TYPES_EXTEND_ORIGIN_TRIAL[f_type]
+  ship_stage_type = STAGE_TYPES_SHIPPING[f_type]
+
+
+  stage_info: StageTemplateInfo = {
+    'proto_stages': [],
+    'dt_stages': [],
+    'ot_stages': [],
+    'extension_stages': [],
+    'ship_stages': [],
+    # Note if any milestones that can be displayed are seen while organizing.
+    # This is used to check if rendering the milestone table is needed.
+    'should_render_mstone_table': False,
+    # Note if any intent URLs are seen while organizing.
+    # This is used to check if rendering the table is needed.
+    'should_render_intents': False
+  }
+
+  for s in Stage.query(Stage.feature_id == id):
+    # Stage info is not needed if it's not the correct stage type
+    # or has no milestones specified.
+    if (s.stage_type != proto_stage_type and
+        s.stage_type != dt_stage_type and
+        s.stage_type != ot_stage_type and
+        s.stage_type != extension_stage_type and
+        s.stage_type != ship_stage_type):
+      continue
+
+    # If an intent thread is present in any stage,
+    # we should render the intents template.
+    stage_info['should_render_intents'] = (
```

Comment:
You could use `stage_info['should_render_intents'] |= s.intent_thread_url is not None`
but it seems more clear as:
```
if s.intent_thread_url is not None:
  stage_info['should_render_intents'] = True
```

Reply:
Yes, much clearer. 🙂